pygfx/objects/_base.py

```diff
@@ -294,63 +257,65 @@ def receive_shadow(self, value):
-        """The child objects of this wold object (read-only tuple).
-        Use ``.add()`` and ``.remove()`` to change this list.
-        """
-        return tuple(self._children)
+    def parent(self) -> "WorldObject":
+        if self._parent is None:
+            return None
+        else:
+            return self._parent()
+
+    @parent.setter
+    def parent(self, value: "WorldObject"):
+        if value is None:
+            self._parent = None
+        else:
+            self._parent = weakref.ref(value)
 
     def add(self, *objects, before=None):
         """Adds object as child of this object. Any number of
         objects may be added. Any current parent on an object passed
         in here will be removed, since an object can have at most one
         parent.
-        If ``before`` argument is given (and present in children), then
-        the items are inserted before the given element.
+        If ``before`` argument is given, then the items are inserted before the
+        given element.
+
         """
-        idx = len(self._children)
-        if before:
-            try:
-                idx = self._children.index(before)
-            except ValueError:
-                pass
+
+        obj: WorldObject
         for obj in objects:
-            assert isinstance(obj, WorldObject)
-            # orphan if needed
-            if obj._parent_ref is not None:
-                obj._parent_ref().remove(obj)
-            # attach to scene graph
-            obj._parent_ref = weakref.ref(self)
-            self._children.insert(idx, obj)
-            idx += 1
-            # flag world matrix as dirty
-            obj._matrix_world_dirty = True
-        return self
+            # added for backwards compatibility
+            # (I feel like we should raise instead if the object is already part
+            # of the scene graph)
+            if obj.parent is not None:
+                obj.parent.remove(obj)
+
+            if before is not None:
+                idx = self.children.index(before)
+            else:
+                idx = len(self.children)
+
+            obj.parent = self
+            self.children.insert(idx, obj)
 
     def remove(self, *objects):
-        """Removes object as child of this object. Any number of objects may be removed.
-        If a given object is not a child, it is ignored.
-        """
+        """Removes object as child of this object. Any number of objects may be removed."""
+
+        obj: WorldObject
         for obj in objects:
-            try:
-                self._children.remove(obj)
-                obj._parent_ref = None
-            except ValueError:
-                pass
-        return self
+            obj.parent = None
+            self.children.remove(obj)
 
     def clear(self):
         """Removes all children."""
-        for child in self._children:
-            child._parent_ref = None
-        self._children.clear()
+        self.remove(*self.children)
```

Comment:
although this is nice and concise, it's much more expensive implementation than the old one.

```suggestion
        for child in self.children:
            child.parent = None
        self.children.clear()
```

Reply:
Sure. I mainly want to share the codepath for cleaning up a child after it was removed from the scene graph.

Reply:
I can relate to that, but I think it's worth the tradeoff for performance here.

Reply:
I created a dedicated `WorldObject._reset_parent` function that I now call from `clear` and `remove` respectively.